ledger/src/leader_schedule/vote_keyed.rs

```diff
@@ -0,0 +1,251 @@
+}
+
+impl LeaderSchedule {
+    // Note: passing in zero vote accounts will cause a panic.
+    pub fn new(
+        vote_accounts_map: &VoteAccountsHashMap,
+        epoch: Epoch,
+        len: u64,
+        repeat: u64,
+    ) -> Self {
+        let keyed_stakes: Vec<_> = vote_accounts_map
+            .iter()
+            .map(|(vote_pubkey, (stake, _account))| (vote_pubkey, *stake))
+            .collect();
+        let vote_keyed_slot_leaders = stake_weighted_slot_leaders(keyed_stakes, epoch, len, repeat);
+        Self::new_from_schedule(vote_keyed_slot_leaders, vote_accounts_map)
+    }
+
+    fn new_from_schedule(
+        vote_keyed_slot_leaders: Vec<Pubkey>,
+        vote_accounts_map: &VoteAccountsHashMap,
+    ) -> Self {
+        struct SlotLeaderInfo<'a> {
+            vote_account_address: &'a Pubkey,
+            validator_identity_address: &'a Pubkey,
+        }
+
+        let default_pubkey = Pubkey::default();
+        let mut current_slot_leader_info = SlotLeaderInfo {
+            vote_account_address: &default_pubkey,
+            validator_identity_address: &default_pubkey,
+        };
+
+        let slot_leaders: Vec<Pubkey> = vote_keyed_slot_leaders
+            .iter()
+            .map(|vote_account_address| {
+                if vote_account_address != current_slot_leader_info.vote_account_address {
+                    let validator_identity_address = vote_accounts_map
+                        .get(vote_account_address)
+                        .expect("vote account must be in vote_accounts_map")
+                        .1
+                        .node_pubkey();
+                    current_slot_leader_info = SlotLeaderInfo {
+                        vote_account_address,
+                        validator_identity_address,
+                    };
+                }
+                *current_slot_leader_info.validator_identity_address
+            })
+            .collect();
+
+        Self {
+            vote_keyed_slot_leaders,
+            identity_keyed_leader_schedule: IdentityKeyedLeaderSchedule::new_from_schedule(
+                slot_leaders,
+            ),
+        }
+    }
+}
+
+impl LeaderScheduleVariant for LeaderSchedule {
+    fn get_slot_leaders(&self) -> &[Pubkey] {
+        self.identity_keyed_leader_schedule.get_slot_leaders()
+    }
+
+    fn get_leader_slots_map(&self) -> &HashMap<Pubkey, Arc<Vec<usize>>> {
+        self.identity_keyed_leader_schedule.get_leader_slots_map()
+    }
+
+    fn is_vote_keyed(&self) -> bool {
+        true
+    }
+
+    fn get_vote_key_at_slot_index(&self, index: usize) -> Option<&Pubkey> {
+        let slot_vote_addresses = &self.vote_keyed_slot_leaders;
+        Some(&slot_vote_addresses[index % slot_vote_addresses.len()])
+    }
+}
+
+impl Index<u64> for LeaderSchedule {
+    type Output = Pubkey;
+    fn index(&self, index: u64) -> &Pubkey {
+        &self.get_slot_leaders()[index as usize % self.num_slots()]
```

Comment:
i know this is existing behavior, but very unintuitive to silently wrap an index request :face_with_head_bandage:

Reply:
Yeah would be happy to remove the `Index` impl entirely in a follow up